BxlAsLib.txt

```diff
@@ -0,0 +1,377 @@
+
+- BXLLib is applicable to many customers, not only to Office, e.g., WDG can benefit from BXLLib.
+
+Spike
+-----
+- Use BuildXL DScript frontend to evaluate specs and generate specific fragments.
+  - Description:
+    - This approach targets only performance.
+    - In metabuild, besides generating DScript spec, the pip also evaluates the spec (plus the SDK) and generates a fragment.
+    - Instead of using a full-blown BuildXL, we refactor BuildXL frontend to create so-called BXLScriptAnalyzer
+      that is used for evaluation and for generating fragment.
+    - The generated fragments are then fed as inputs to product build.
+    - Performance can be gained because:
+      - The metabuild pip generates a smaller spec that can be evaluated fast (to be proven).
+      - The metabuild pip can be distributed. 
+
+  - Pros:
+    - It's a shortcut and requires moderate efforts.
+    - We can (hopefully) deliver performance without changing DScript SDK, without changing static NMake, without
+      waiting for definite API.
+    - Existing DScript SDK can still be used.
+    - The shipment of BuildXL engine and its DScript SDK doesn't change.
+
+  - Cons:
+    - This may be applicable to Office where their project slices are kind of isolated from one another.
+    - Stitching fragments is non-trivial due to service and IPC pips:
+      - [In progress] Need to unify service and IPC pips:
+        - Instead of having one drop service pip for the whole graph, we now have N drop service pips, where
+          N is the number of generated fragments.
+        - IPC pips are tied to specific service pips through monikers. We need to mess around with IPC pips
+          command-line arguments to unify service and IPC pips.
+
+API
+---
+- Basic API: must be small and not prone to binary incompatibility.
+  
+  static class FragmentFactory 
+  {
+      /* Fragment's name can be thought of as module name */
+      public IGraphFragment InitGraphFragment(string name);
+  }
+
+  interface IGraphFragment 
+  {
+      void AddProcess(ProcessData processData);
+      void AddWriteFile(WriteFileData writeFileData);
+      void AddCopyFile(CopyFileData copyFileData);
+      void AddSealDirectory(SealDirectoryData sealDirectory);
+      ...
+
+      AbsolutePath P(string path);
+      FileArtifact F(AbsolutePath path, int rewriteCount);
+      DirectoryArtifact D(AbsolutePath path);
+  }
+
+- Some utilites need to be exposed for optimization:
+  - AbsolutePath, 
+  - FileArtifact(WithAttaributs), 
+  - DirectoryArtifact,
+  - PathTable
+  - PipData/PipDataBuilder:
+    - PipData builder is needed to construct PipData.
+    - Need to be moved to utilities.
+    - Some renderer that refers to FileContentInfo of BuildXL.Storage needs to use plain string.
+    - Perhaps create a restricted version of PipDataBuilder that user can use easily.
+  - To support reuse, some classes/enums in BuildXL.Pips needs to be moved to a smaller DLL (may be called

```

Comment:
>BuildXL.Pip [](start = 44, length = 11)

You probalby mean BuildXL.Utilities...

pips has the classes and the base graph.
PathTable and family are in Utilities

Reply:
Note if we can't refactor those out into a small reusable library, It could just be a facade assembly with type forwarders to BuildXL.Utiltiies.

---
In reply to: [304077365](https://github.com/microsoft/BuildXL/pull/603#discussion_r304077365) [](ancestors = 304077365)